R/ingest.R

```diff
@@ -0,0 +1,248 @@
+        opts <- utils::modifyList(opts, list(streamFormat="Csv"))
+    }
+    else body <- readLines(src)
+
+    ingest_uri <- httr::parse_url(database$server)
+    ingest_uri$path <- file.path("v1/rest/ingest", database$database, dest_table)
+    ingest_uri$query <- opts
+
+    headers <- httr::add_headers(Authorization=paste("Bearer", validate_token(ingestion_token)))
+
+    res <- httr::POST(ingest_uri, headers, body=body, encode="raw")
+
+    http_status_handler <- match.arg(http_status_handler)
+    if(http_status_handler == "pass")
+        return(res)
+
+    cont <- httr::content(res, simplifyVector=TRUE)
+    handler <- get(paste0(http_status_handler, "_for_status"), getNamespace("httr"))
+    handler(res, make_error_message(cont))
+    cont
+}
+
+
+ingest_indirect <- function(database, src, dest_table, staging_container=NULL, ...)
+{
+    if(!requireNamespace("AzureStor"))
+        stop("AzureStor package must be installed to do indirect ingestion", call.=FALSE)
+
+    opts <- utils::modifyList(list(...), list(
+        key=staging_container$endpoint$key,
+        token=staging_container$endpoint$token,
+        sas=staging_container$endpoint$sas))
+
+    if(is.data.frame(src))
+    {
+        utils::write.table(src, textConnection("con", "w"), row.names=FALSE, col.names=FALSE, sep=",")
+        src <- textConnection(con, "r")
+        opts <- utils::modifyList(opts, list(streamFormat="Csv"))
+    }
+
+    if(inherits(staging_container, "blob_container"))
+    {
+        uploadfunc <- get("upload_blob", getNamespace("AzureStor"))
+        uploadfunc(staging_container, src, dest_table)
+
+        url <- httr::parse_url(staging_container$endpoint$url)
+        url$path <- file.path(staging_container$name, dest_table)
+
+        do.call(ingest_blob, c(list(database, httr::build_url(url), dest_table), opts))
+    }
+    else if(inherits(staging_container, "adls_filesystem"))
+    {
+        uploadfunc <- get("upload_adls_file", getNamespace("AzureStor"))
+        uploadfunc(staging_container, src, dest_table)
+
+        url <- httr::parse_url(staging_container$endpoint$url)
+        url$scheme <- "abfss"
+        url$username <- staging_container$name
+        url$path <- dest_table
+
+        do.call(ingest_adls2, c(list(database, httr::build_url(url), dest_table), opts))
+    }
+    else stop("Unsupported staging container type", call.=FALSE)
+}
+
+
+ingest_inline <- function(database, src, dest_table, ...)
+{
+    prop_list <- get_ingestion_properties(...)
+
+    if(is.data.frame(src))
+    {
+        con <- textConnection(NULL, "w")
+        on.exit(close(con))
+        utils::write.table(src, con, row.names=FALSE, col.names=FALSE, sep=",")
+        records <- textConnectionValue(con)
+    }
+    else records <- readLines(src)
+
+    cmd <- paste(".ingest inline into table",
+        escape(ident(dest_table)),
+        prop_list,
+        "<|\n",
+        paste0(sapply(records, escape), collapse="\n"))
```

Comment:
To give an example:
```r
> ingest_inline(ingestdb, iris, 'iristestingest')
> res <- run_query(ingestdb, "iristestingest")
> head(res)
  sepal_length sepal_width petal_length petal_width species
1           NA         3.5          1.4         0.2  setosa
2           NA          NA           NA          NA        
3           NA         3.0          1.4         0.2  setosa
4           NA          NA           NA          NA        
5           NA         3.2          1.3         0.2  setosa
6           NA          NA           NA          NA        
> 
```

Reply:
Actually, I don't think the records in `ingest_inline` should be escaped. This is just verbatim text, right?

> Data is the data content to ingest. Unless otherwise modified by the ingestion properties, this content is parsed as CSV. Note that unlike most control commands and queries, the text of the Data part of the command does not have to follow the syntactic conventions of the language (e.g., whitespace characters are important, the // combination is not treated as a comment, etc.)

https://kusto.azurewebsites.net/docs/management/data-ingestion/ingest-inline.html

Reply:
It's a bit far-fetched in the case of R and Kusto, but I can imagine an application that asks a user to enter or upload CSV data and the user injects a malicious command like ".drop table" inside the CSV text. It sounds like the Kusto service already escapes the CSV data,